lib/Doctrine/Common/Cache/Psr6/DoctrineProvider.php

```diff
@@ -66,6 +75,31 @@ protected function doFetch($id)
         return $item->isHit() ? $item->get() : false;
     }
 
+    /**
+     * @inheritDoc
+     */
+    protected function doFetchMultiple(array $keys): array
+    {
+        $encodedKeys = array_map('rawurlencode', $keys);
+
+        $items = $this->pool->getItems($encodedKeys);
+        if ($items instanceof Traversable) {
+            $items = iterator_to_array($items);
+        }
+
+        // Sorting items by keys
+        $items = array_replace(array_flip($encodedKeys), $items);
```

Comment:
This code is basically looping over keys 4 times to build the final results. I think a foreach loop working in imperative way will be faster:

```
$result = [];

foreach ($keys as $key)  {
    // To be checked whether redoing the rawurlencode is slower or faster than using a single `array_combine` to create a mapping of keys to encoded keys, or comparing that to building both the list and the map with a loop instead of the array_map
    $item = $items[rawurlencode($key)];

    if (!$item->isHit()) {
        continue;
    }

    $results[$key] = $item->get();
}
```

Reply:
Changed in 4155612